### ISSUE TITLE
ClassCastException in JpaHibernateConnectionProvider with spring 5.3

### ISSUE THREAD

**Issue by 7erg** [javers/javers#1034]
Exception:
java.lang.ClassCastException: class com.sun.proxy.$Proxy113 cannot be cast to class org.hibernate.internal.SessionImpl (com.sun.proxy.$Proxy113 and org.hibernate.internal.SessionImpl are in unnamed module of loader 'app')
	at org.javers.spring.jpa.JpaHibernateConnectionProvider.getConnection(JpaHibernateConnectionProvider.java:22)

Here
`(SessionImpl)entityManager.unwrap(Session.class);`

`entityManager.unwrap(Session.class)` now returns Proxy instead of SessionImpl instance, because of changes in HibernateJpaVendorAdapter: entityManagerInterface changed from HibernateEntityManager (deprecated) to Session.

Probable fix:  `(SessionImpl)entityManager.unwrap(SessionImpl.class);`

**Comment by qanik1307** [comment]
Probably related: java.lang.ClassCastException: class com.sun.proxy.$Proxy185 cannot be cast to class org.hibernate.engine.spi.SessionFactoryImplementor (com.sun.proxy.$Proxy185 and org.hibernate.engine.spi.SessionFactoryImplementor are in unnamed module of loader 'app')
	at org.javers.spring.boot.sql.JaversSqlAutoConfiguration.javersSqlDialectName(JaversSqlAutoConfiguration.java:55)

**Comment by temofey1989** [comment]
Unfortunately `javersSqlDialectName` bean has no `@ConditionalOnMissingBean` annotation. So, it hard to override this bean.

Also, the same problem with casting in the `JpaHibernateConnectionProvider`.

I'm using:
- Spring Boot 2.4.0
- Javers 5.13.2

The workaround for me is:
- exclude Javers autoconfiguration
  ```
  @SpringBootApplication(
      exclude = [JaversSqlAutoConfiguration::class]
  )
  ```
- define a custom configuration with overrides:
  ```
  @Configuration
  class Config : JaversSqlAutoConfiguration() {

      @PersistenceContext
      lateinit var entityManager: EntityManager

      @Autowired
      lateinit var entityManagerFactory: EntityManagerFactory

      @Bean
      override fun javersSqlDialectName(): DialectName {
          val implementor = entityManagerFactory.unwrap(SessionFactoryImplementor::class.java)
          return DialectMapper().map(implementor.jdbcServices.dialect)
      }

      @Bean
      override fun jpaConnectionProvider(): ConnectionProvider {
          return ConnectionProvider {
              val session = (entityManager.unwrap(Session::class.java) as EntityManagerProxy).targetEntityManager as SessionImpl
              session.connection()
          }
      }
  }
  ```
I hope this would be useful for someone... :smiley:

**Comment by bartoszwalacik** [comment]
Which version of Hibernate do you use?

**Comment by 7erg** [comment]
> Which version of Hibernate do you use?

5.4.23.Final

**Comment by bartoszwalacik** [comment]
https://github.com/javers/javers/pull/1035

**Comment by bartoszwalacik** [comment]
fixed in 5.14.0
